### PR TITLE
Support enableCompression in workflow and job configs

### LINKED ISSUE #884: Allow users to set enableCompression for workflow and job configs
See title.

### PR DESCRIPTION
**Pull request by narendly**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Fixes #884 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Previously, there was no way for users to set enableCompression in workflows and jobs configs. This allows them to set that field in the Builder.

### Tests

- [x] The following is the result of the "mvn test" command on the appropriate module:

```
Failed tests: 
  TestEnableCompression.testEnableCompressionResource:108 expected:<true> but was:<false>
  TestHelixAdminCli.testDeactivateCluster:604 » Helix There are still LEADER in ...

Tests run: 1087, Failures: 2, Errors: 0, Skipped: 1

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:08 h
[INFO] Finished at: 2020-03-11T23:48:14-07:00
[INFO] ------------------------------------------------------------------------
```

```
Results :

Tests run: 11, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  47.491 s
[INFO] Finished at: 2020-03-12T09:48:20-07:00
[INFO] ------------------------------------------------------------------------
```

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

**Comment by narendly** [comment]
> Do we want to just add this field or support a builder can carry over all the fields from JobConfig?

Just this field only because we want to curb the way user creates workflows and jobs now. Users should ideally use PropertyStore to store user-generated custom fields.

**Comment by pkuwm** [comment]
> > Do we want to just add this field or support a builder can carry over all the fields from JobConfig?
> 
> Just this field only because we want to curb the way user creates workflows and jobs now. Users should ideally use PropertyStore to store user-generated custom fields.

Can we add a log to tell users their properties are ignored if any? I think that would be helpful for the end users.

File: helix-core/src/main/java/org/apache/helix/task/JobConfig.java [hunk 1]
```diff
@@ -553,7 +556,11 @@ public static Builder fromMap(Map<String, String> cfg) {
       }
       if (cfg.containsKey(JobConfigProperty.RebalanceRunningTask.name())) {
         b.setRebalanceRunningTask(
-            Boolean.valueOf(cfg.get(JobConfigProperty.RebalanceRunningTask.name())));
+            Boolean.parseBoolean(cfg.get(JobConfigProperty.RebalanceRunningTask.name())));
+      }
+      if (cfg.containsKey(ZNRecord.ENABLE_COMPRESSION_BOOLEAN_FIELD)) {
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/task/JobConfig.java]: It is redundant to check `containsKey`: if the value is not set, parseBoolean still returns false.
Unless the default value enableCompression is true, but I don't think this should be the case.

File: helix-core/src/main/java/org/apache/helix/task/WorkflowConfig.java [hunk 2]
```diff
@@ -552,6 +559,11 @@ public Builder setConfigMap(Map<String, String> cfg) {
         }
       }
 
+      if (cfg.containsKey(ZNRecord.ENABLE_COMPRESSION_BOOLEAN_FIELD)) {
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/task/WorkflowConfig.java]: No need to check containsKey, unless the default value enableCompression is true.

**Comment by narendly** [comment]
> > > Do we want to just add this field or support a builder can carry over all the fields from JobConfig?
> > 
> > 
> > Just this field only because we want to curb the way user creates workflows and jobs now. Users should ideally use PropertyStore to store user-generated custom fields.
> 
> Can we add a log to tell users their properties are ignored if any? I think that would be helpful for the end users.

I don't think adding logs is necessary. Logs are helpful for leaving historical information that can't be found anywhere else. In this case, however, users can easily check whether their fields ever made it to the znode by inspecting the znode.

Also, we want to limit the kind of fields we allow users to include in workflow and job config znodes by having them use Builders.

The biggest worry of all is again for heavy users of Task Framework, this might pollute the log.

File: helix-core/src/main/java/org/apache/helix/task/JobConfig.java [hunk 3]
```diff
@@ -553,8 +556,11 @@ public static Builder fromMap(Map<String, String> cfg) {
       }
       if (cfg.containsKey(JobConfigProperty.RebalanceRunningTask.name())) {
         b.setRebalanceRunningTask(
-            Boolean.valueOf(cfg.get(JobConfigProperty.RebalanceRunningTask.name())));
+            Boolean.parseBoolean(cfg.get(JobConfigProperty.RebalanceRunningTask.name())));
       }
+      // If not set, parseBoolean returns false for null
+      b.setEnableCompression(
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/task/JobConfig.java]: Then we should do a check here, right? If get null, we will get exception for parse boolean.

Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/task/JobConfig.java]: @dasahcc 

This is the implementation for parseBoolean() if the comment wasn't clear:

```
    public static boolean parseBoolean(String s) {
        return ((s != null) && s.equalsIgnoreCase("true"));
    }
```
Do you see why the null check is not necessary?

Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/task/JobConfig.java]:
```
    public static boolean parseBoolean(String s) {
        return ((s != null) && s.equalsIgnoreCase("true"));
    }
```

Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/task/JobConfig.java]: OK. The null should be final. But it make this entry explicitly in ZNRecord, right? Usually, our check is if the entry is not in ZNRecord, we use default value. But we dont carry the default value and explicitly set in ZNRecord to reduce the size of data set.

In this case, it is better we follow our convention: do the check when we carry it over.

Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/task/JobConfig.java]: Okay I see what you mean. I will revert to the original pattern.

**Pull request merged**